### PR TITLE
[SPARK-49441][ML] `StringIndexer` sort arrays in executors

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
`StringIndexer` sort arrays in executors

### Why are the changes needed?
Existing implementation collect all arrays at driver, and then sort them with a `ForkJoinPool` via `ThreadUtils.parmap`.

This PR aims to move sorting from driver to executors.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
CI and manually test:
```
import org.apache.spark.ml.feature.StringIndexer

val numCol = 300

val data = (0 to 10000).map { i =>
  (i, 100 * i)
}
var df = data.toDF("id", "label0")
(1 to numCol).foreach { idx =>
  df = df.withColumn(s"label$idx", col("label0") + 1)
}
val inputCols = (0 to numCol).map(i => s"label$i").toArray
val outputCols = (0 to numCol).map(i => s"labelIndex$i").toArray
df.cache()
df.count()

Seq.range(0, 10).foreach(i => new StringIndexer().setInputCols(inputCols).setOutputCols(outputCols).setStringOrderType("alphabetDesc").fit(df)) // warm up


val t0 = System.nanoTime()
Seq.range(0, 10).foreach(i => new StringIndexer().setInputCols(inputCols).setOutputCols(outputCols).setStringOrderType("alphabetDesc").fit(df))
val t1 = System.nanoTime()
println("Elapsed time: " + (t1 - t0) / 1000000000.0 + "s")
```

master:
```
Elapsed time: 12.181134084s
```

this PR:
```
Elapsed time: 11.432259833s
```

The performance is similar, the main change is that the sort happens in executors.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala [hunk 1]
```diff
@@ -197,53 +196,43 @@ class StringIndexer @Since("1.4.0") (
...
+    val result = Array.fill(numCols)(Array.empty[String])
     dataset.select(posexplode(array(selectedCols: _*)).as(Seq("index", "value")))
       .where(col("value").isNotNull)
       .groupBy("index", "value")
-      .agg(count(lit(1)).as("count"))
+      .agg(countCol.as("count"))
       .groupBy("index")
-      .agg(collect_list(struct("value", "count")))
+      .agg(collect_list(struct("count", "value")).alias("array"))
+      .repartition(8, col("index"))
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:214]: Out of curiosity, why 8 here? can we avoid the repartition()? I don't immediately see why we need it

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:214]: to be consistent with original parallelism:
```
    ThreadUtils.parmap(orgStrings, "sortingStringLabels", 8) { counts =>
      counts.toSeq.sortWith(sortFunc).map(_._1).toArray
    }.toArray
```

on second thought, I think we can remove it

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:214]: in the benchmark, removing it won't hurt the performance:
```
Elapsed time: 10.855109542s
```

Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala]: I get it, just seems like a repartition() is potentially expensive. I don't know if there was much significance to 8, other than it being a bit parallel. Here we already have parallelism in place and it's a small operation. I'd imagine you can remove it for simplicity, even if it isn't faster at small scale.

**Review by HyukjinKwon — APPROVED** [review]

**Review by WeichenXu123 — APPROVED** [review]
LGTM!

**Comment by zhengruifeng** [comment]
thanks all, merged to master

**Pull request closed without merging**